liblwgeom/lwgeom.c

```diff
@@ -2505,37 +2505,34 @@ lwgeom_is_trajectory(const LWGEOM *geom)
 	return lwline_is_trajectory((LWLINE*)geom);
 }
 
-static uint8_t
-bits_for_precision(int32_t significant_digits)
-{
-	int32_t bits_needed = ceil(significant_digits / log10(2));
-
-	if (bits_needed > 52)
-	{
-		return 52;
-	}
-	else if (bits_needed < 1)
-	{
-		return 1;
-	}
-
-	return bits_needed;
-}
+#define STATIC_ASSERT(COND,MSG) typedef char static_assertion_##MSG[(COND)?1:-1]
+STATIC_ASSERT(sizeof(double) == sizeof(uint64_t),this_should_be_true);
 
 static double trim_preserve_decimal_digits(double d, int32_t decimal_digits)
 {
-	if (d == 0)
-		return 0;
-
-	int digits_left_of_decimal = (int) (1 + log10(fabs(d)));
-	uint8_t bits_needed = bits_for_precision(decimal_digits + digits_left_of_decimal);
-	uint64_t mask = 0xffffffffffffffffULL << (52 - bits_needed);
 	uint64_t dint = 0;
-	size_t dsz = sizeof(d) < sizeof(dint) ? sizeof(d) : sizeof(dint);
-
-	memcpy(&dint, &d, dsz);
+	memcpy(&dint, &d, sizeof(double));
```

Comment:
I believe that copy-less recommended way of dealing with the type punning in C is using a two-part struct. Kinda like here: https://github.com/postgis/postgis/blob/master/postgis/gserialized_gist_2d.c#L458

Reply:
memcpy() is also fine (cf https://stackoverflow.com/a/53181506). It is heavily used in GDAL and optimizers do not do memory copy, but move things between registries (at least on x86). 
I believe what you actually recommend is undefined behaviour (also most compilers do sane things in this situation): cf https://stackoverflow.com/a/69297979

Reply:
It's UB in C++ but not in C: https://stackoverflow.com/a/25672839. 

I checked disasm and it looks good there indeed, thanks.
![image](https://user-images.githubusercontent.com/810638/209709638-93a23886-0527-44e1-8161-21ab8f73ff16.png)

Reply:
> It's UB in C++ but not in C

interesting... thanks